lib/THC/generic/THCTensorMath.cu

```diff
@@ -122,17 +122,71 @@ void THCTensor_(catArray)(THCState *state, THCTensor *result,
   THCTensor_(resize)(state, result, size, NULL);
   THLongStorage_free(size);
 
-  offset = 0;
-  for (j = 0; j < numInputs; j++)
-  {
-    long dimSize = dimension < THCTensor_(nDimension)(state, inputs[j])
-			       ? THCTensor_(size)(state, inputs[j], dimension)
-			       : 1;
-    THCTensor *nt = THCTensor_(newWithTensor)(state, result);
-    THCTensor_(narrow)(state, nt, NULL, dimension, offset, dimSize);
-    THCTensor_(copy)(state, nt, inputs[j]);
-    THCTensor_(free)(state, nt);
-    offset += dimSize;
+  // We parallelize the copy if all 3 conditions pass:
+  //
+  // 1. There is more than one input tensor
+  // 2. All input tensors can use 32-bit indexing
+  // 3. All input tensors are on the same device
+
+  if (numInputs > 1 &&
+      TensorUtils<THCTensor>::all32BitIndexable(state, inputs, numInputs) &&
+      TensorUtils<THCTensor>::allSameDevice(state, inputs, numInputs)) {
+
+    // First, define a TensorInfo for the result tensor to pass to the kernel
+    TensorInfo<real, unsigned int> rst = getTensorInfo<THCTensor, unsigned int>(state, result);
+
+    // The basic strategy is as follows: We batch copies of the remaining input
+    // tensors by passing a struct to the kernel containing TensorInfos for a fixed
+    // number of inputs, along with information to handle the narrow, copy within the kernel call
+    index = 0;
+    for (i = 0; i < numInputs; i += CAT_ARRAY_KERNEL_BATCH_SIZE) {
+    /* for (i = 0; i < 1; ++i) { */
+      CatArrayKernelParam<real> param;
+      cohortMax = 0;
+      for (j = 0; j < CAT_ARRAY_KERNEL_BATCH_SIZE && (i+j) < numInputs; ++j) {
+      /* for (j = 0; j < 1 && (i+j) < numInputs; ++j) { */
+        param.inputs[j] = getTensorInfo<THCTensor, unsigned int>(state, inputs[i+j]);
+        param.offsets[j] = index;
+        param.dimSizes[j] = dimension < THCTensor_(nDimension)(state, inputs[i+j])
+               ? THCTensor_(size)(state, inputs[i+j], dimension)
+               : 1;
+        index += param.dimSizes[j];
+        param.nElements[j] = THCTensor_(nElement)(state, inputs[i+j]);
+        cohortMax = cohortMax > param.nElements[j] ? cohortMax : param.nElements[j];
+      }
+      param.count = j;
+
```

Comment:
Consider using constant memory to store `param` so it more arrays can be copied in a single kernel launch?